kotlinpoet-metadata-specs/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/KotlinPoetMetadataSpecs.kt

```diff
@@ -206,7 +206,13 @@ private fun ImmutableKmClass.toTypeSpec(
 ): TypeSpec {
   val classTypeParamsResolver = typeParameters.toTypeParamsResolver()
 
-  val simpleName = name.substringAfterLast(if (isInline) "/" else ".")
+  // package/of/class/MyClass.InnerClass
+  // Sometimes it's package/of/class/MyClass$InnerClass
+  val simpleName = name.substringAfterLast(charArrayOf(
+      '/', // Drop the package name, e.g. "package/of/class/"
+      '.', // Drop any enclosing classes, e.g. "MyClass."
+      '$'  // Drop any enclosing classes, e.g. "MyClass$"
```

Comment:
What happens for
```kotlin
class `Foo$Bar`
```
?

Reply:
It becomes `com/squareup/kotlinpoet/metadata/specs/test/Foo$Bar`

I'm not really sure what to do about this though :|. Any suggestions?

Reply:
There will be an InnerClass attribute in the classfile when it's an actually-nested type. Not sure if that's exposed to you or not from wherever you're reading.

Reply:
Hacking on this - We can solve 95% of this case by just asking the source `Class`/`TypeElement`/`KClass` for its simple name, and the metadata gives us the simple name when we iterate over subtypes. The remaining 5% would just be cases where someone calls `ImmutableKmClass.toTypeSpec`, in which case we can accept an optional `simpleName` that defaults to the current implementation as a "best guess" approach. We can tune that best guess approach via element handler as well in the future to check the existence of that type and whatnot.

I have a mostly working implementation of this locally, but due to how Elements tries to canonicalize names, I think we should push to a separate PR  pending its API cleanup #776, where we could give it a hint of what the known simple name is.